File: src/main/java/net/Chidoziealways/everythingjapanese/block/entity/ModBlockEntities.java
package net.Chidoziealways.everythingjapanese.block.entity;

import net.Chidoziealways.everythingjapanese.EverythingJapanese;
import net.Chidoziealways.everythingjapanese.block.ModBlocks;
import net.Chidoziealways.everythingjapanese.block.entity.custom.GrowthChamberBlockEntity;
import net.Chidoziealways.everythingjapanese.block.entity.custom.PedestalBlockEntity;
import net.minecraft.core.Registry;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import net.minecraftforge.registries.RegistryObject;

import java.util.Set;

public class ModBlockEntities {
    public static final DeferredRegister<BlockEntityType<?>> BLOCK_ENTITIES =
            DeferredRegister.create(ForgeRegistries.BLOCK_ENTITY_TYPES, EverythingJapanese.MOD_ID);

    public static final RegistryObject<BlockEntityType<PedestalBlockEntity>> PEDESTAL_BE =
            BLOCK_ENTITIES.register("pedestal_be", () -> {
              BlockEntityType<PedestalBlockEntity> type = new BlockEntityType<>(PedestalBlockEntity::new, Set.of(ModBlocks.PEDESTAL.get()));
              return Registry.register(BuiltInRegistries.BLOCK_ENTITY_TYPE,
                      ResourceLocation.fromNamespaceAndPath(EverythingJapanese.MOD_ID, "pedestal_be"), type);
            });

    public static final RegistryObject<BlockEntityType<GrowthChamberBlockEntity>> GROWTH_CHAMBER_BE =
            BLOCK_ENTITIES.register("growth_chamber_be", () -> {
                BlockEntityType<GrowthChamberBlockEntity> type = new BlockEntityType<>(GrowthChamberBlockEntity::new, Set.of(ModBlocks.GROWTH_CHAMBER.get()));
                return Registry.register(BuiltInRegistries.BLOCK_ENTITY_TYPE,
                        ResourceLocation.fromNamespaceAndPath(EverythingJapanese.MOD_ID, "growth_chamber_be"), type);
            });

    public static void register(IEventBus eventBus) {
        BLOCK_ENTITIES.register(eventBus);
    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/block/entity/ModBlockEntities.kt
package net.Chidoziealways.everythingjapanese.entity

import net.Chidoziealways.everythingjapanese.MOD_ID
import net.Chidoziealways.everythingjapanese.block.ModBlocks
import net.Chidoziealways.everythingjapanese.block.entity.custom.ByoubuBlockEntity
import net.Chidoziealways.everythingjapanese.block.entity.custom.FusumaDoorBlockEntity
import net.Chidoziealways.everythingjapanese.block.entity.custom.ShojiDoorBlockEntity
import net.Chidoziealways.everythingjapanese.entity.custom.GrowthChamberBlockEntity
import net.Chidoziealways.everythingjapanese.entity.custom.PedestalBlockEntity
import net.minecraft.core.BlockPos
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.entity.BlockEntityType.BlockEntitySupplier
import net.minecraft.world.level.block.state.BlockState
import net.neoforged.bus.api.IEventBus
import net.neoforged.neoforge.registries.DeferredRegister
import thedarkcolour.kotlinforforge.neoforge.forge.getValue

object ModBlockEntities {
    val BLOCK_ENTITIES = DeferredRegister.create(Registries.BLOCK_ENTITY_TYPE, MOD_ID)

    val SHOJI_DOOR_BE by BLOCK_ENTITIES.register("shoji_door_be") { ->
        BlockEntityType(BlockEntitySupplier { pPos: BlockPos, pBlockState: BlockState ->
            ShojiDoorBlockEntity(
                pPos,
                pBlockState) },
            mutableSetOf<Block>(ModBlocks.SHOJI_DOOR))
    }

    val FUSUMA_DOOR_BE by BLOCK_ENTITIES.register("fusuma_door_be") { ->
        BlockEntityType(BlockEntitySupplier { pPos: BlockPos, pBlockState: BlockState ->
            FusumaDoorBlockEntity(
                pPos,
                pBlockState) },
            mutableSetOf<Block>(ModBlocks.FUSUMA_DOOR))
    }

    val BYOUBU_BE by BLOCK_ENTITIES.register("byoubu_be") { ->
        BlockEntityType(BlockEntitySupplier { pos, state ->
            ByoubuBlockEntity(
                pos,
                state
            ) },
            mutableSetOf<Block>(ModBlocks.BYOUBU))
    }

    val PEDESTAL_BE by
        BLOCK_ENTITIES.register("pedestal_be") { ->
                BlockEntityType(BlockEntitySupplier { pPos: BlockPos, pBlockState: BlockState ->
                    PedestalBlockEntity(
                        pPos,
                        pBlockState
                    ) },
                    mutableSetOf<Block?>(ModBlocks.PEDESTAL))
        }

    val GROWTH_CHAMBER_BE by BLOCK_ENTITIES.register("growth_chamber_be") { ->
        BlockEntityType(
            BlockEntitySupplier { pPos: BlockPos, pState: BlockState ->
                GrowthChamberBlockEntity(
                    pPos,
                    pState
                ) },
            mutableSetOf<Block?>(ModBlocks.GROWTH_CHAMBER)
        )
    }

    fun register(eventBus: IEventBus) {
        BLOCK_ENTITIES.register(eventBus)
    }
}
